File: fontes/backend/java/fenix/src/main/java/com/t2ti/fenix/model/cadastros/Pessoa.java
/*****************************************************************
Title: T2Ti ERP Fenix
Description: Model relacionado a tabela []
License: Apache2
Copyright: Copyright 2020 (C) William Marx

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.

      The author may be contacted at:
          marxwilliamf@gmail.com

@Author: William Felipe Marx (marxwilliamf@gmail.com)
@version: 1.0.0
*****************************************************************/

package com.t2ti.fenix.model.cadastros;

import java.io.Serializable;
import java.util.Set;

import javax.persistence.CascadeType;
import javax.persistence.Entity;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.NamedQuery;
import javax.persistence.OneToMany;
import javax.persistence.OneToOne;

/**
 * The persistent class for the pessoa database table.
 * 
 */
@Entity
@NamedQuery(name="Pessoa.findAll", query="SELECT p FROM Pessoa p")
public class Pessoa implements Serializable {
	private static final long serialVersionUID = 1L;

	@Id
	@GeneratedValue(strategy = GenerationType.IDENTITY)
	private Integer id;
	private String nome;
	private String tipo;
	private String site;
	private String email;
	private String cliente;
	private String colaborador;
	private String fornecedor;
	private String transportadora;
	private String contador;

	//bi-directional many-to-one association to PessoaContato
	@OneToMany(mappedBy="pessoa", cascade = CascadeType.ALL, orphanRemoval = true)
	private Set<PessoaContato> listaPessoaContato;

	//bi-directional many-to-one association to PessoaEndereco
	@OneToMany(mappedBy="pessoa", cascade = CascadeType.ALL, orphanRemoval = true)
	private Set<PessoaEndereco> listaPessoaEndereco;

	//bi-directional many-to-one association to PessoaFisica
	@OneToOne(mappedBy="pessoa", cascade = CascadeType.ALL, orphanRemoval = true)
	private PessoaFisica pessoaFisica;
	
	@OneToOne(mappedBy="pessoa", cascade = CascadeType.ALL, orphanRemoval = true)
	private PessoaJuridica pessoaJuridica;

	//bi-directional many-to-one association to PessoaTelefone
	@OneToMany(mappedBy="pessoa", cascade = CascadeType.ALL, orphanRemoval = true)
	private Set<PessoaTelefone> listaPessoaTelefone;

	public Pessoa() {
	}

	public Integer getId() {
		return this.id;
	}

	public void setId(Integer id) {
		this.id = id;
	}

	public String getCliente() {
		return this.cliente;
	}

	public void setCliente(String cliente) {
		this.cliente = cliente;
	}

	public String getColaborador() {
		return this.colaborador;
	}

	public void setColaborador(String colaborador) {
		this.colaborador = colaborador;
	}

	public String getContador() {
		return this.contador;
	}

	public void setContador(String contador) {
		this.contador = contador;
	}

	public String getEmail() {
		return this.email;
	}

	public void setEmail(String email) {
		this.email = email;
	}

	public String getFornecedor() {
		return this.fornecedor;
	}

	public void setFornecedor(String fornecedor) {
		this.fornecedor = fornecedor;
	}

	public String getNome() {
		return this.nome;
	}

	public void setNome(String nome) {
		this.nome = nome;
	}

	public String getSite() {
		return this.site;
	}

	public void setSite(String site) {
		this.site = site;
	}

	public String getTipo() {
		return this.tipo;
	}

	public void setTipo(String tipo) {
		this.tipo = tipo;
	}

	public String getTransportadora() {
		return this.transportadora;
	}

	public void setTransportadora(String transportadora) {
		this.transportadora = transportadora;
	}

	public Set<PessoaContato> getListaPessoaContato() {
		return this.listaPessoaContato;
	}

	public void setListaPessoaContato(Set<PessoaContato> listaPessoaContato) {
		this.listaPessoaContato = listaPessoaContato;
		for(PessoaContato pessoaContato : this.listaPessoaContato) {
			pessoaContato.setPessoa(this);
		}
	}

	public Set<PessoaEndereco> getListaPessoaEndereco() {
		return this.listaPessoaEndereco;
	}

	public void setListaPessoaEndereco(Set<PessoaEndereco> listaPessoaEndereco) {
		this.listaPessoaEndereco = listaPessoaEndereco;
		for(PessoaEndereco pessoaEndereco:this.listaPessoaEndereco) {
			pessoaEndereco.setPessoa(this);
		}
	}

	public PessoaFisica getPessoaFisica() {
		return this.pessoaFisica;
	}

	public void setPessoaFisica(PessoaFisica pessoaFisica) {
		this.pessoaFisica = pessoaFisica;
		this.pessoaFisica.setPessoa(this);
	}

	public PessoaJuridica getPessoaJuridica() {
		return this.pessoaJuridica;
	}

	public void setPessoaJuridica(PessoaJuridica pessoaJuridica) {
		this.pessoaJuridica = pessoaJuridica;
		this.pessoaJuridica.setPessoa(this);
	}

	public Set<PessoaTelefone> getListaPessoaTelefone() {
		return this.listaPessoaTelefone;
	}

	public void setListaPessoaTelefone(Set<PessoaTelefone> listaPessoaTelefone) {
		this.listaPessoaTelefone = listaPessoaTelefone;
		for(PessoaTelefone pessoaTelefone : this.listaPessoaTelefone) {
			pessoaTelefone.setPessoa(this);
		}
	}

}

File: fontes/frontend/Flutter/provider/fenix/lib/src/model/cadastros/pessoa.dart
import 'dart:convert';

import 'package:fenix/src/model/cadastros/pessoa_contato.dart';
import 'package:fenix/src/model/cadastros/pessoa_endereco.dart';
import 'package:fenix/src/model/cadastros/pessoa_fisica.dart';
import 'package:fenix/src/model/cadastros/pessoa_juridica.dart';
import 'package:fenix/src/model/cadastros/pessoa_telefone.dart';

class Pessoa {
  int id;
  String nome;
  String tipo;
  String site;
  String email;
  String cliente;
  String fornecedor;
  String transportadora;
  String colaborador;
  String contador;
  PessoaFisica pessoaFisica;
  PessoaJuridica pessoaJuridica;
  List<PessoaTelefone> listaPessoaTelefone = [];
  List<PessoaEndereco> listaPessoaEndereco = [];
  List<PessoaContato> listaPessoaContato = [];

  Pessoa(
      {this.id,
      this.nome,
      this.tipo = 'Física', //valor padrçao para o preenchimento do DropDownButton
      this.site,
      this.email,
      this.cliente,
      this.fornecedor,
      this.transportadora,
      this.colaborador,
      this.contador,
      this.pessoaFisica,
      this.pessoaJuridica,
      this.listaPessoaTelefone,
      this.listaPessoaEndereco,
      this.listaPessoaContato});

  static List<String> campos = <String>[
    'id',
    'nome',
    'tipo',
    'site',
    'email',
    'cliente',
    'fornecedor',
    'transportadora',
    'colaborador',
    'contador'
  ];
  static List<String> colunas = <String>[
    'Id',
    'Nome',
    'Tipo',
    'Site',
    'Email',
    'Cliente',
    'Fornecedor',
    'Transportadora',
    'Colaborador',
    'Contador'
  ];

  Pessoa.fromJson(Map<String, dynamic> jsonDados) {
    id = jsonDados['id'];
    nome = jsonDados['nome'];
    tipo = (jsonDados['tipo'] ?? 'F') == 'F' ? 'Física' : 'Jurídica'; //DropDownButton  // Domains
    site = jsonDados['site'];
    email = jsonDados['email'];
    cliente = jsonDados['cliente'] == 'S' ? 'Sim' : 'Não'; //DropDownButton
    fornecedor = jsonDados['fornecedor'] == 'S' ? 'Sim' : 'Não'; //DropDownButton
    transportadora = jsonDados['transportadora'] == 'S' ? 'Sim' : 'Não'; //DropDownButton
    colaborador = jsonDados['colaborador'] == 'S' ? 'Sim' : 'Não'; //DropDownButton
    contador = jsonDados['contador'] == 'S' ? 'Sim' : 'Não'; //DropDownButton
    pessoaFisica = jsonDados['pessoaFisica'] == null ? null : new PessoaFisica.fromJson(jsonDados['pessoaFisica']);
    pessoaJuridica = jsonDados['pessoaJuridica'] == null ? null : new PessoaJuridica.fromJson(jsonDados['pessoaJuridica']);
    listaPessoaTelefone = (jsonDados['listaPessoaTelefone'] as Iterable)?.map((m) => PessoaTelefone.fromJson(m))?.toList() ?? [];
    listaPessoaEndereco = (jsonDados['listaPessoaEndereco'] as Iterable)?.map((m) => PessoaEndereco.fromJson(m))?.toList() ?? [];
    listaPessoaContato = (jsonDados['listaPessoaContato'] as Iterable)?.map((m) => PessoaContato.fromJson(m))?.toList() ?? [];
  }

  Map<String, dynamic> get toJson {
    Map<String, dynamic> jsonDados = Map<String, dynamic>();
    jsonDados['id'] = this.id;
    jsonDados['nome'] = this.nome;
    jsonDados['tipo'] = (this.tipo ?? 'Física') == 'Física' ? 'F' : 'J'; //DropDownButton
    jsonDados['site'] = this.site;
    jsonDados['email'] = this.email;
    jsonDados['cliente'] = this.cliente == 'Sim' ? 'S' : 'N'; //DropDownButton
    jsonDados['fornecedor'] = this.fornecedor == 'Sim' ? 'S' : 'N'; //DropDownButton
    jsonDados['transportadora'] = this.transportadora == 'Sim' ? 'S' : 'N'; //DropDownButton
    jsonDados['colaborador'] = this.colaborador == 'Sim' ? 'S' : 'N'; //DropDownButton
    jsonDados['contador'] = this.contador == 'Sim' ? 'S' : 'N'; //DropDownButton
    jsonDados['pessoaFisica'] = this.pessoaFisica == null ? null : this.pessoaFisica.toJson;
    jsonDados['pessoaJuridica'] = this.pessoaJuridica == null ? null : this.pessoaJuridica.toJson;

    var listaPessoaTelefoneLocal = [];
    for (PessoaTelefone pessoaTelefone in this.listaPessoaTelefone ?? []) {
      listaPessoaTelefoneLocal.add(pessoaTelefone.toJson);
    }
    jsonDados['listaPessoaTelefone'] = listaPessoaTelefoneLocal;

    var listaPessoaEnderecoLocal = [];
    for (PessoaEndereco pessoaEndereco in this.listaPessoaEndereco ?? []) {
      listaPessoaEnderecoLocal.add(pessoaEndereco.toJson);
    }
    jsonDados['listaPessoaEndereco'] = listaPessoaEnderecoLocal;
    var listaPessoaContatoLocal = [];
    for (PessoaContato pessoaContato in this.listaPessoaContato ?? []) {
      listaPessoaContatoLocal.add(pessoaContato.toJson);
    }
    jsonDados['listaPessoaContato'] = listaPessoaContatoLocal;

    return jsonDados;
  }
}

String pessoaEncodeJson(Pessoa pessoa) {
  final jsonDados = pessoa.toJson;
  return json.encode(jsonDados);
}
